File: pythonbetter/dtopattern2.py
class UserDTO:
    def __init__(self, id: int, name: str, email: str):
        self.id = id
        self.name = name
        self.email = email

    @classmethod
    def from_dict(cls, data: dict):
        return cls(
            id=data.get('id', 0),
            name=data.get('name', ''),
            email=data.get('email', '')
        )

    def to_dict(self) -> dict:
        return vars(self)

# Przykład użycia
data = {
    "id": 2,
    "name": "Alice",
    "email": "alice@example.com"
}

user_dto = UserDTO.from_dict(data)
print(user_dto.to_dict())
# {'id': 2, 'name': 'Alice', 'email': 'alice@example.com'}

File: somephp/dtopattern2.php
<?php
class UserDTO implements JsonSerializable {
    public int $id;
    public string $name;
    public string $email;

    public function __construct(array $data) {
        $this->id = $data['id'] ?? 0;
        $this->name = $data['name'] ?? '';
        $this->email = $data['email'] ?? '';
    }

    public function jsonSerialize(): array {
        return get_object_vars($this);
    }
}

// Przykład użycia
$data = [
    'id' => 1,
    'name' => 'Jane Doe',
    'email' => 'jane.doe@example.com'
];

$userDTO = new UserDTO($data);

// Serializacja do JSON
echo json_encode($userDTO);

// {"id":1,"name":"Jane Doe","email":"jane.doe@example.com"}